madara/crates/client/sync/src/l2.rs

```diff
@@ -224,6 +232,69 @@ async fn l2_pending_block_task(
     Ok(())
 }
 
+async fn l2_highest_block_fetch(
+    provider: Arc<GatewayProvider>,
+    mut ctx: ServiceContext,
+    sync_status_provider: Arc<SyncStatusProvider>,
+) -> anyhow::Result<()> {
+    tracing::debug!("Start highest block poll");
+
+    let mut interval = tokio::time::interval(Duration::from_secs(10));
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+    // Track consecutive failures to prevent endless loops
+    let mut consecutive_failures = 0;
+
+    // Keep running until cancelled
+    while ctx.run_until_cancelled(interval.tick()).await.is_some() {
+        tracing::debug!("Getting highest block...");
+
+        // Try to get the latest block
+        match provider.get_block(BlockId::Tag(BlockTag::Latest)).await {
+            Ok(block) => {
+                match &block {
+                    ProviderBlockPendingMaybe::NonPending(block_info) => {
+                        // For a regular block, we have both number and hash
+                        let block_number = block_info.block_number;
+                        let block_hash = block_info.block_hash;
+
+                        // Update the sync status provider with the highest block info
+                        sync_status_provider.set_highest_block_num(block_number).await;
+                        sync_status_provider.set_highest_block_hash(block_hash).await;
+
+                        tracing::debug!("Updated highest block: number={}, hash={}", block_number, block_hash);
+
+                        // Reset failure counter after successful processing
+                        consecutive_failures = 0;
+                    }
+                    _ => {
+                        tracing::error!("Got unexpected block type from provider that wasn't handled");
+                        consecutive_failures += 1;
+                        tokio::time::sleep(Duration::from_secs(HIGHEST_BLOCK_FETCH_ERROR_DELAY_SECS)).await;
+                    }
+                }
+            }
+            Err(e) => {
+                tracing::error!("Error getting latest block: {:?}", e);
+                consecutive_failures += 1;
+                tokio::time::sleep(Duration::from_secs(HIGHEST_BLOCK_FETCH_ERROR_DELAY_SECS)).await;
+            }
+        }
+
+        // If we've had too many consecutive failures, break out of the loop
+        if consecutive_failures >= MAX_HIGHEST_BLOCK_FETCH_FAILURES {
+            let error_msg = format!(
+                "Received unexpected block types or errors for {MAX_HIGHEST_BLOCK_FETCH_FAILURES} consecutive attempts"
+            );
```

Comment:
Maybe you meant this?
```suggestion
            let error_msg = format!(
                "Received unexpected block types or errors for {} consecutive attempts", MAX_HIGHEST_BLOCK_FETCH_FAILURES
            );
```

Reply:
These are equivalent :)